docs/man/man3/ElementsCollectionCache.3
.TH "ElementsCollectionCache" 3 "Thu Aug 27 2020" "Version 0.8-dev" "QElectroTech" \" -*- nroff -*-
.ad l
.nh
.SH NAME
ElementsCollectionCache
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <elementscollectioncache\&.h>\fP
.PP
Inherits QObject\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBElementsCollectionCache\fP (const QString &database_path, QObject *=nullptr)"
.br
.ti -1c
.RI "\fB~ElementsCollectionCache\fP () override"
.br
.ti -1c
.RI "void \fBsetLocale\fP (const QString &)"
.br
.ti -1c
.RI "QString \fBlocale\fP () const"
.br
.ti -1c
.RI "bool \fBsetPixmapStorageFormat\fP (const QString &)"
.br
.ti -1c
.RI "QString \fBpixmapStorageFormat\fP () const"
.br
.ti -1c
.RI "bool \fBfetchElement\fP (\fBElementsLocation\fP &location)"
.br
.RI "\fBElementsCollectionCache::fetchElement\fP Retrieve the data for a given element, using the cache if available, filling it otherwise\&. Data are then available through \fBpixmap()\fP and \fBname()\fP methods\&. "
.ti -1c
.RI "QString \fBname\fP () const"
.br
.ti -1c
.RI "QPixmap \fBpixmap\fP () const"
.br
.ti -1c
.RI "bool \fBfetchData\fP (const \fBElementsLocation\fP &)"
.br
.ti -1c
.RI "bool \fBfetchNameFromCache\fP (const QString &path, const QUuid &uuid)"
.br
.RI "\fBElementsCollectionCache::fetchNameFromCache\fP Retrieve the name for an element, given its path and uuid The value is then available through the \fBname()\fP method\&. "
.ti -1c
.RI "bool \fBfetchPixmapFromCache\fP (const QString &path, const QUuid &uuid)"
.br
.RI "\fBElementsCollectionCache::fetchPixmapFromCache\fP Retrieve the pixmap for an element, given its path and uuid\&. It is then available through the \fBpixmap()\fP method\&. "
.ti -1c
.RI "bool \fBcacheName\fP (const QString &path, const QUuid &uuid=QUuid::createUuid())"
.br
.RI "\fBElementsCollectionCache::cacheName\fP Cache the current (i\&.e\&. last retrieved) name The cache entry will use the locale set via \fBsetLocale()\fP\&. "
.ti -1c
.RI "bool \fBcachePixmap\fP (const QString &path, const QUuid &uuid=QUuid::createUuid())"
.br
.RI "\fBElementsCollectionCache::cachePixmap\fP Cache the current (i\&.e\&. last retrieved) pixmap\&. "
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "QSqlDatabase \fBcache_db_\fP"
.br
.RI "Object providing access to the SQLite database this cache relies on\&. "
.ti -1c
.RI "QSqlQuery * \fBselect_name_\fP"
.br
.RI "Prepared statement to fetch names from the cache\&. "
.ti -1c
.RI "QSqlQuery * \fBselect_pixmap_\fP"
.br
.RI "Prepared statement to fetch pixmaps from the cache\&. "
.ti -1c
.RI "QSqlQuery * \fBinsert_name_\fP"
.br
.RI "Prepared statement to insert names into the cache\&. "
.ti -1c
.RI "QSqlQuery * \fBinsert_pixmap_\fP"
.br
.RI "Prepared statement to insert pixmaps into the cache\&. "
.ti -1c
.RI "QString \fBlocale_\fP"
.br
.RI "Locale to be used when dealing with names\&. "
.ti -1c
.RI "QString \fBpixmap_storage_format_\fP"
.br
.RI "Storage format for cached pixmaps\&. "
.ti -1c
.RI "QString \fBcurrent_name_\fP"
.br
.RI "Last name fetched\&. "
.ti -1c
.RI "QPixmap \fBcurrent_pixmap_\fP"
.br
.RI "Last pixmap fetched\&. "
.in -1c
.SH "Detailed Description"
.PP 
This class implements a SQLite cache for data related to elements collections, mainly names and pixmaps\&. This avoids the cost of parsing XML definitions of elements and building full CustomElement objects when (re)loading the elements panel\&. 
.SH "Constructor & Destructor Documentation"
.PP 
.SS "ElementsCollectionCache::ElementsCollectionCache (const QString & database_path, QObject * parent = \fCnullptr\fP)"
Construct a cache for elements collections\&. 
.PP
\fBParameters\fP
.RS 4
\fIdatabase_path\fP Path of the SQLite database to open\&. 
.br
\fIparent\fP Parent QObject 
.RE
.PP

.SS "ElementsCollectionCache::~ElementsCollectionCache ()\fC [override]\fP"
Destructor 
.SH "Member Function Documentation"
.PP 
.SS "bool ElementsCollectionCache::cacheName (const QString & path, const QUuid & uuid = \fCQUuid::createUuid()\fP)"

.PP
\fBElementsCollectionCache::cacheName\fP Cache the current (i\&.e\&. last retrieved) name The cache entry will use the locale set via \fBsetLocale()\fP\&. 
.PP
\fBParameters\fP
.RS 4
\fIpath\fP : \fBElement\fP path (as obtained using \fBElementsLocation::toString()\fP) 
.br
\fIuuid\fP :\fBElement\fP uuid 
.RE
.PP
\fBReturns\fP
.RS 4
True if the caching succeeded, false otherwise\&. 
.RE
.PP
\fBSee also\fP
.RS 4
\fBname()\fP 
.RE
.PP

.SS "bool ElementsCollectionCache::cachePixmap (const QString & path, const QUuid & uuid = \fCQUuid::createUuid()\fP)"

.PP
\fBElementsCollectionCache::cachePixmap\fP Cache the current (i\&.e\&. last retrieved) pixmap\&. 
.PP
\fBParameters\fP
.RS 4
\fIpath\fP : \fBElement\fP path (as obtained using \fBElementsLocation::toString()\fP) 
.br
\fIuuid\fP : \fBElement\fP uuid 
.RE
.PP
\fBReturns\fP
.RS 4
True if the caching succeeded, false otherwise\&. 
.RE
.PP
\fBSee also\fP
.RS 4
\fBpixmap()\fP 
.RE
.PP

.SS "bool ElementsCollectionCache::fetchData (const \fBElementsLocation\fP & location)"
Retrieve the data by building the full CustomElement object matching the given location, without using the cache\&. Data are then available through \fBpixmap()\fP and \fBname()\fP methods\&. 
.PP
\fBParameters\fP
.RS 4
\fIlocation\fP Location of a given \fBElement\fP\&. 
.RE
.PP
\fBReturns\fP
.RS 4
True if the retrieval succeeded, false otherwise\&. 
.RE
.PP

.SS "bool ElementsCollectionCache::fetchElement (\fBElementsLocation\fP & location)"

.PP
\fBElementsCollectionCache::fetchElement\fP Retrieve the data for a given element, using the cache if available, filling it otherwise\&. Data are then available through \fBpixmap()\fP and \fBname()\fP methods\&. 
.PP
\fBParameters\fP
.RS 4
\fIlocation\fP The definition of an element\&. 
.RE
.PP
\fBSee also\fP
.RS 4
\fBpixmap()\fP 
.PP
\fBname()\fP 
.RE
.PP
\fBReturns\fP
.RS 4
True if the retrieval succeeded, false otherwise\&. 
.RE
.PP

.SS "bool ElementsCollectionCache::fetchNameFromCache (const QString & path, const QUuid & uuid)"

.PP
\fBElementsCollectionCache::fetchNameFromCache\fP Retrieve the name for an element, given its path and uuid The value is then available through the \fBname()\fP method\&. 
.PP
\fBParameters\fP
.RS 4
\fIpath\fP : \fBElement\fP path (as obtained using \fBElementsLocation::toString()\fP) 
.br
\fIuuid\fP : \fBElement\fP uuid 
.RE
.PP
\fBReturns\fP
.RS 4
True if the retrieval succeeded, false otherwise\&. 
.RE
.PP

.SS "bool ElementsCollectionCache::fetchPixmapFromCache (const QString & path, const QUuid & uuid)"

.PP
\fBElementsCollectionCache::fetchPixmapFromCache\fP Retrieve the pixmap for an element, given its path and uuid\&. It is then available through the \fBpixmap()\fP method\&. 
.PP
\fBParameters\fP
.RS 4
\fIpath\fP : \fBElement\fP path (as obtained using \fBElementsLocation::toString()\fP) 
.br
\fIuuid\fP : \fBElement\fP uuid 
.RE
.PP
\fBReturns\fP
.RS 4
True if the retrieval succeeded, false otherwise\&. 
.RE
.PP

.SS "QString ElementsCollectionCache::locale () const"

.PP
\fBReturns\fP
.RS 4
The locale to be used when dealing with names\&. 
.RE
.PP

.SS "QString ElementsCollectionCache::name () const"

.PP
\fBReturns\fP
.RS 4
The last name fetched through \fBfetchElement()\fP\&. 
.RE
.PP

.SS "QPixmap ElementsCollectionCache::pixmap () const"

.PP
\fBReturns\fP
.RS 4
The last pixmap fetched through \fBfetchElement()\fP\&. 
.RE
.PP

.SS "QString ElementsCollectionCache::pixmapStorageFormat () const"

.PP
\fBReturns\fP
.RS 4
the pixmap storage format\&. Default is 'PNG' 
.RE
.PP
\fBSee also\fP
.RS 4
\fBsetPixmapStorageFormat()\fP 
.RE
.PP

.SS "void ElementsCollectionCache::setLocale (const QString & locale)"
Define the locale to be used when dealing with names\&. 
.PP
\fBParameters\fP
.RS 4
\fIlocale\fP New locale to be used\&. 
.RE
.PP

.SS "bool ElementsCollectionCache::setPixmapStorageFormat (const QString & format)"
Define the storage format for the pixmaps within the SQLite database\&. See Qt's QPixmap documentation for more information\&. 
.PP
\fBParameters\fP
.RS 4
\fIformat\fP The new pixmap storage format\&. 
.RE
.PP
\fBReturns\fP
.RS 4
True if the format change was accepted, false otherwise\&. 
.RE
.PP

.SH "Member Data Documentation"
.PP 
.SS "QSqlDatabase ElementsCollectionCache::cache_db_\fC [private]\fP"

.PP
Object providing access to the SQLite database this cache relies on\&. 
.SS "QString ElementsCollectionCache::current_name_\fC [private]\fP"

.PP
Last name fetched\&. 
.SS "QPixmap ElementsCollectionCache::current_pixmap_\fC [private]\fP"

.PP
Last pixmap fetched\&. 
.SS "QSqlQuery* ElementsCollectionCache::insert_name_\fC [private]\fP"

.PP
Prepared statement to insert names into the cache\&. 
.SS "QSqlQuery* ElementsCollectionCache::insert_pixmap_\fC [private]\fP"

.PP
Prepared statement to insert pixmaps into the cache\&. 
.SS "QString ElementsCollectionCache::locale_\fC [private]\fP"

.PP
Locale to be used when dealing with names\&. 
.SS "QString ElementsCollectionCache::pixmap_storage_format_\fC [private]\fP"

.PP
Storage format for cached pixmaps\&. 
.SS "QSqlQuery* ElementsCollectionCache::select_name_\fC [private]\fP"

.PP
Prepared statement to fetch names from the cache\&. 
.SS "QSqlQuery* ElementsCollectionCache::select_pixmap_\fC [private]\fP"

.PP
Prepared statement to fetch pixmaps from the cache\&. 

.SH "Author"
.PP 
Generated automatically by Doxygen for QElectroTech from the source code\&.


sources/elementscollectioncache.cpp
/*
	Copyright 2006-2020 The QElectroTech Team
	This file is part of QElectroTech.
	
	QElectroTech is free software: you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation, either version 2 of the License, or
	(at your option) any later version.
	
	QElectroTech is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.
	
	You should have received a copy of the GNU General Public License
	along with QElectroTech.  If not, see <http://www.gnu.org/licenses/>.
*/
#include "elementscollectioncache.h"
#include "factory/elementfactory.h"
#include "element.h"
#include "qet.h"

#include <QImageWriter>
#include <QSqlQuery>
#include <QSqlError>

/**
	Construct a cache for elements collections.
	@param database_path Path of the SQLite database to open.
	@param parent Parent QObject
*/
ElementsCollectionCache::ElementsCollectionCache(const QString &database_path,
						 QObject *parent) :
	QObject(parent),
	locale_("en"),
	pixmap_storage_format_("PNG")
{
	// initialize the cache SQLite database
	static int cache_instances = 0;
	QString connection_name = QString("ElementsCollectionCache-%1").arg(
				cache_instances++);
	cache_db_ = QSqlDatabase::addDatabase("QSQLITE", connection_name);
	cache_db_.setDatabaseName(database_path);

	if (!cache_db_.open())
		qDebug() << "Unable to open the SQLite database "
			 << database_path
			 << " as "
			 << connection_name
			 << ": "
			 << cache_db_.lastError();
	else
	{
		cache_db_.exec("PRAGMA temp_store = MEMORY");
		cache_db_.exec("PRAGMA journal_mode = MEMORY");
		cache_db_.exec("PRAGMA page_size = 4096");
		cache_db_.exec("PRAGMA cache_size = 16384");
		cache_db_.exec("PRAGMA locking_mode = EXCLUSIVE");
		cache_db_.exec("PRAGMA synchronous = OFF");
#pragma message("@TODO the tables could already exist, handle that case.")
		//@TODO the tables could already exist, handle that case.
		cache_db_.exec("CREATE TABLE names"
					   "("
					   "path VARCHAR(512) NOT NULL,"
					   "locale VARCHAR(2) NOT NULL,"
					   "uuid VARCHAR(512) NOT NULL,"
					   "name VARCHAR(128),"
					   "PRIMARY KEY(path, locale)"
					   ");");

		cache_db_.exec("CREATE TABLE pixmaps"
					   "("
					   "path VARCHAR(512) NOT NULL UNIQUE,"
					   "uuid VARCHAR(512) NOT NULL,"
					   "pixmap BLOB, PRIMARY KEY(path),"
					   "FOREIGN KEY(path) REFERENCES names (path) ON DELETE CASCADE);");

		// prepare queries
		select_name_   = new QSqlQuery(cache_db_);
		select_pixmap_ = new QSqlQuery(cache_db_);
		insert_name_   = new QSqlQuery(cache_db_);
		insert_pixmap_ = new QSqlQuery(cache_db_);
		select_name_   -> prepare("SELECT name FROM names WHERE path = :path AND locale = :locale AND uuid = :uuid");
		select_pixmap_ -> prepare("SELECT pixmap FROM pixmaps WHERE path = :path AND uuid = :uuid");
		insert_name_   -> prepare("REPLACE INTO names (path, locale, uuid, name) VALUES (:path, :locale, :uuid, :name)");
		insert_pixmap_ -> prepare("REPLACE INTO pixmaps (path, uuid, pixmap) VALUES (:path, :uuid, :pixmap)");
	}
}

/**
	Destructor
*/
ElementsCollectionCache::~ElementsCollectionCache() {
	delete select_name_;
	delete select_pixmap_;
	delete insert_name_;
	delete insert_pixmap_;
	cache_db_.close();
}

/**
	Define the locale to be used when dealing with names.
	@param locale New locale to be used.
*/
void ElementsCollectionCache::setLocale(const QString &locale) {
	locale_ = locale;
}

/**
	@return The locale to be used when dealing with names.
*/
QString ElementsCollectionCache::locale() const {
	return(locale_);
}

/**
	Define the storage format for the pixmaps within the SQLite database. See
	Qt's QPixmap documentation for more information.
	@param format The new pixmap storage format.
	@return True if the format change was accepted, false otherwise.
*/
bool ElementsCollectionCache::setPixmapStorageFormat(const QString &format) {
	if (QImageWriter::supportedImageFormats().contains(format.toLatin1())) {
		pixmap_storage_format_= format;
		return(true);
	}
	return(false);
}

/**
	@return the pixmap storage format. Default is "PNG"
	@see setPixmapStorageFormat()
*/
QString ElementsCollectionCache::pixmapStorageFormat() const {
	return(pixmap_storage_format_);
}

/**
 * @brief ElementsCollectionCache::fetchElement
 * Retrieve the data for a given element, using the cache if available,
 * filling it otherwise. Data are then available through pixmap() and name() methods.
 * @param location The definition of an element.
 * @see pixmap()
 * @see name()
 * @return True if the retrieval succeeded, false otherwise.
 */
bool ElementsCollectionCache::fetchElement(ElementsLocation &location)
{
		// can we use the cache with this element?
	bool use_cache = cache_db_.isOpen() && !location.isProject();

		// attempt to fetch the element name from the cache database
	if (!use_cache) {
		return(fetchData(location));
	}
	else
	{
		auto uuid = location.uuid();
		QString element_path = location.toString();
		bool got_name   = fetchNameFromCache(element_path, uuid);
		bool got_pixmap = fetchPixmapFromCache(element_path, uuid);

		if (got_name && got_pixmap) {
			return(true);
		}

		if (fetchData(location))
		{
			cacheName(element_path, uuid);
			cachePixmap(element_path, uuid);
		}
		return(true);
	}
}

/**
	@return The last name fetched through fetchElement().
*/
QString ElementsCollectionCache::name() const {
	return(current_name_);
}

/**
	@return The last pixmap fetched through fetchElement().
*/
QPixmap ElementsCollectionCache::pixmap() const {
	return(current_pixmap_);
}

/**
	Retrieve the data by building the full CustomElement object matching the
	given location, without using the cache. Data are then available through
	pixmap() and name() methods.
	@param Location Location of a given Element.
	@return True if the retrieval succeeded, false otherwise.
*/
bool ElementsCollectionCache::fetchData(const ElementsLocation &location) {
	int state;
	Element *custom_elmt = ElementFactory::Instance() -> createElement(location, nullptr, &state);
	if (state) {
		qDebug() << "ElementsCollectionCache::fetchData() : Le chargement du composant" << qPrintable(location.toString()) << "a echoue avec le code d'erreur" << state;
	} else {
		current_name_   = custom_elmt -> name();
		current_pixmap_ = custom_elmt -> pixmap();
	}
	delete custom_elmt;
	return(!state);
}

/**
 * @brief ElementsCollectionCache::fetchNameFromCache
 * Retrieve the name for an element, given its path and uuid
 * The value is then available through the name() method.
 * @param path : Element path (as obtained using ElementsLocation::toString())
 * @param uuid : Element uuid
 * @return True if the retrieval succeeded, false otherwise.
 */
bool ElementsCollectionCache::fetchNameFromCache(const QString &path, const QUuid &uuid)
{
	select_name_ -> bindValue(":path", path);
	select_name_ -> bindValue(":locale", locale_);
	select_name_ -> bindValue(":uuid", uuid.toString());
	if (select_name_ -> exec())
	{
		if (select_name_ -> first())
		{
			current_name_ = select_name_ -> value(0).toString();
			select_name_ -> finish();
			return(true);
		}
	}
	else
		qDebug() << "select_name_->exec() failed";

	return(false);
}

/**
 * @brief ElementsCollectionCache::fetchPixmapFromCache
 * Retrieve the pixmap for an element, given its path and uuid.
 * It is then available through the pixmap() method.
 * @param path : Element path (as obtained using ElementsLocation::toString())
 * @param uuid : Element uuid
 * @return True if the retrieval succeeded, false otherwise.
 */
bool ElementsCollectionCache::fetchPixmapFromCache(const QString &path, const QUuid &uuid)
{
	select_pixmap_ -> bindValue(":path", path);
	select_pixmap_ -> bindValue(":uuid", uuid.toString());
	if (select_pixmap_ -> exec())
	{
		if (select_pixmap_ -> first())
		{
			QByteArray ba = select_pixmap_ -> value(0).toByteArray();
			// avoid returning always the same pixmap (i.e. same cacheKey())
			current_pixmap_.detach();
			current_pixmap_.loadFromData(ba, qPrintable(pixmap_storage_format_));
			select_pixmap_ -> finish();
		}
		return(true);
	}
	else
		qDebug() << "select_pixmap_->exec() failed";

	return(false);
}

/**
 * @brief ElementsCollectionCache::cacheName
 * Cache the current (i.e. last retrieved) name The cache entry will use the locale set via setLocale().
 * @param path : Element path (as obtained using ElementsLocation::toString())
 * @param uuid :Element uuid
 * @return True if the caching succeeded, false otherwise.
 * @see name()
 */
bool ElementsCollectionCache::cacheName(const QString &path, const QUuid &uuid)
{
	insert_name_ -> bindValue(":path",   path);
	insert_name_ -> bindValue(":locale", locale_);
	insert_name_ -> bindValue(":uuid",  uuid.toString());
	insert_name_ -> bindValue(":name",   current_name_);
	if (!insert_name_ -> exec())
	{
		qDebug() << cache_db_.lastError();
		return(false);
	}
	return(true);
}

/**
 * @brief ElementsCollectionCache::cachePixmap
 * Cache the current (i.e. last retrieved) pixmap
 * @param path : Element path (as obtained using ElementsLocation::toString())
 * @param uuid : Element uuid
 * @return True if the caching succeeded, false otherwise.
 * @see pixmap()
 */
bool ElementsCollectionCache::cachePixmap(const QString &path, const QUuid &uuid)
{
	QByteArray ba;
	QBuffer buffer(&ba);
	buffer.open(QIODevice::WriteOnly);
	current_pixmap_.save(&buffer, qPrintable(pixmap_storage_format_));
	insert_pixmap_ -> bindValue(":path", path);
	insert_pixmap_ -> bindValue(":uuid", uuid.toString());
	insert_pixmap_ -> bindValue(":pixmap", QVariant(ba));
	if (!insert_pixmap_->exec())
	{
		qDebug() << cache_db_.lastError();
		return(false);
	}
	return(true);
}
